src/plugins/scopes/ScopesApi.js

```diff
@@ -0,0 +1,474 @@
+import ArrowFunctionExpression from '../../elements/types/ArrowFunctionExpression';
+import VariableDeclarator from '../../elements/types/VariableDeclarator';
+import VariableDeclaration from '../../elements/types/VariableDeclaration';
+import AssignmentExpression from '../../elements/types/AssignmentExpression';
+import UpdateExpression from '../../elements/types/UpdateExpression';
+import ForOfStatement from '../../elements/types/ForOfStatement';
+import ForInStatement from '../../elements/types/ForInStatement';
+import MemberExpression from '../../elements/types/MemberExpression';
+import Property from '../../elements/types/Property';
+import MethodDefinition from '../../elements/types/MethodDefinition';
+import ImportDefaultSpecifier from '../../elements/types/ImportDefaultSpecifier';
+import ImportNamespaceSpecifier from '../../elements/types/ImportNamespaceSpecifier';
+import ImportSpecifier from '../../elements/types/ImportSpecifier';
+import ThisExpression from '../../elements/types/ThisExpression';
+import Super from '../../elements/types/Super';
+import CatchClause from '../../elements/types/CatchClause';
+import LabeledStatement from '../../elements/types/LabeledStatement';
+import BreakStatement from '../../elements/types/BreakStatement';
+import ContinueStatement from '../../elements/types/ContinueStatement';
+import ClassExpression from '../../elements/types/ClassExpression';
+import ClassDeclaration from '../../elements/types/ClassDeclaration';
+import JSXIdentifier from '../../elements/types/JSXIdentifier';
+import JSXAttribute from '../../elements/types/JSXAttribute';
+import JSXElement from '../../elements/types/JSXElement';
+import JSXMemberExpression from '../../elements/types/JSXMemberExpression';
+import JSXNamespacedName from '../../elements/types/JSXNamespacedName';
+import {types} from './Definition';
+
+const scopedBlocks = {
+    'ForStatement': true,
+    'ForInStatement': true,
+    'ForOfStatement': true,
+    'SwitchStatement': true,
+    'CatchClause': true
+};
+
+export default class ScopesApi {
+    constructor(program: Program) {
+        program.on('elements-add', (elements: Element[]) => {
+            elements.map(this._addElement, this);
+        });
+        program.on('elements-remove', (elements: Element[]) => {
+            elements.map(this._removeElement, this);
+        });
+        this._scopesMap = new Map();
+        this._program = program;
+        this._addElement(program);
+    }
+
+    _program: Program;
+    _scopesMap: Map<Element, Scope>;
+
+    _addElement(element: Element) {
+        if (element instanceof Node) {
+            let nodes = buildNodeList((element: Node));
+            for (let i = 0; i < nodes.length; i++) {
+                this._addNode(nodes[i]);
+            }
+        }
+    }
+
+    _removeElement(element: Element) {
+        if (element instanceof Node) {
+            let nodes = buildNodeList((element: Node));
```

Comment:
I'm not in favour of using `const` for local `currently-non-changing` variables.